### PR TITLE
Add support for disabling spell checking

### PR DESCRIPTION
**Pull request by SimonBrandner**
Fixes https://github.com/vector-im/element-web/issues/21901
Requires https://github.com/vector-im/element-web/pull/22220
Requires https://github.com/vector-im/element-desktop/pull/379

<hr>

Peek 2022-06-26 11-41

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add support for disabling spell checking ([\#8604](https://github.com/matrix-org/matrix-react-sdk/pull/8604)). Fixes vector-im/element-web#21901.

### CODE REVIEW

**Comment by dbkr** [comment]
Does this need to be exposed on platforms that don't support multi-language spell-checking?

**Comment by t3chguy** [comment]
@SimonBrandner yes but then that means that https://github.com/vector-im/element-web/issues/21901 is not fixed for macOS Desktop users, as there we'll still be passing `spellcheck: true` without a way to disable

File: src/components/views/settings/tabs/user/GeneralUserSettingsTab.tsx [hunk 1]
```diff
@@ -238,10 +244,20 @@ export default class GeneralUserSettingsTab extends React.Component<IProps, ISta
...
-        if (plaf) {
-            plaf.setSpellCheckLanguages(languages);
+    private onSpellCheckEnabledChange = (spellCheckEnabled: boolean): void => {
+        this.setState({ spellCheckEnabled });
+
+        if (spellCheckEnabled && IS_MAC) {
+            // XXX: On macOS the languages are controlled by the system
+            // settings, so we just set something, so that spell check gets
+            // enabled
+            PlatformPeg.get()?.setSpellCheckLanguages(["en_US"]);
```
Comment (turt2live) [inline src/components/views/settings/tabs/user/GeneralUserSettingsTab.tsx:257]: we should probably just turn it onto a simple toggle if we can't support the platform.

**Review by turt2live — COMMENTED** [review]
generally seems fine, and I really wish this was a feature that got a breadcrumb - I didn't know you could change/add dictionaries. I've been out here suffering US English while in Canada.

File: src/components/views/settings/tabs/user/GeneralUserSettingsTab.tsx (hunk 1 above)
Comment (SimonBrandner) [reply src/components/views/settings/tabs/user/GeneralUserSettingsTab.tsx:257]: Currently, the intention is for no spell check languages to be the same as spell check being disabled, so it makes sense to only have a method that sets the spell check languages. 

I'd like to wait to see what Jano thinks of the current behaviour and make changes then, if necessary

**Comment by t3chguy** [comment]
/me will run a full macOS build to test this

**Comment by t3chguy** [comment]
So on macOS it defaulted to on in the UI but spell-check wasn't working. Turning spell-check in the UI off then back on made it work again.

**Comment by SimonBrandner** [comment]
> So on macOS it defaulted to on in the UI but spell-check wasn't working. Turning spell-check in the UI off then back on made it work again.

So I think this might be an unrelated issue where the electron-store value isn't synced with the languages and the actual spell check setting

**Review by janogarcia — COMMENTED** [review]
Had no idea that we supported multi-language spell-checking. On which platforms is it exposed (I either use the Web or macOS version)?

Could you please @SimonBrandner share a couple screenshots so that I can add them to the Settings redesign project? Thanks!

As for the implementation, could we prevent the disable action from clearing the dictionaries preferences?

Ideally, it would disable the spell-checker but would keep the dictionary preferences in case the user want to enable it again. Bundling multiple actions (disable + clear dictionary list) can be unexpected/problematic otherwise when performed without further notice.

**Comment by t3chguy** [comment]
> On which platforms is it exposed (I either use the Web or macOS version)?

The only platforms which support changing your spell check language are Element Desktop on Windows & Linux. Web everywhere it is up to your web browser, Element Desktop macOS it is up to your OS.

**Comment by SimonBrandner** [comment]
@janogarcia, I've updated the PR; below is a video of how the UI behaves now along with a few screenshots. How does this looK?

<hr>

Peek 2022-06-26 11-41

**Review by turt2live — APPROVED** [review]

**Review by janogarcia — COMMENTED** [review]
@SimonBrandner Thanks for updating the behavior as suggested and providing the GIF/screenshots! 👌

One minor request. If the Combobox component supports it, please add a placeholder text "Choose a locale" for the initial/empty state. If it doesn't, please ignore this request, and let's think about improving that component at some point.

**Pull request merged**